lib/events.js

```diff
@@ -1105,33 +1033,48 @@ function on(emitter, event, options) {
       { once: true });
   }
 
-  return iterator;
+  function errorHandler (err) {
+    error = err;
+    if (resume) {
+      resume(Promise.reject(err));
+      resume = null;
+    }
+  }
 
   function abortListener() {
     errorHandler(new AbortError(undefined, { cause: signal?.reason }));
   }
 
   function eventHandler(...args) {
-    const promise = ArrayPrototypeShift(unconsumedPromises);
-    if (promise) {
-      promise.resolve(createIterResult(args, false));
+    if (resume) {
+      resume(args);
+      resume = null;
     } else {
-      unconsumedEvents.push(args);
+      queue.push(args);
     }
   }
 
-  function errorHandler(err) {
-    finished = true;
+  return async function * () {
+    try {
+      while (true) {
+        while (queue.length) {
+          if (error) {
+            throw error;
+          }
+          yield queue.shift();
+        }
 
-    const toError = ArrayPrototypeShift(unconsumedPromises);
+        if (error) {
```

Comment:
this needs a symbol in case someone does something stupid like `throw false`?

Reply:
Fixed